File: examples/python/part-5/16-monitoring-alerting.py
# Illustrative examples for notes/part-5-control-observability/16-monitoring-alerting.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Dict, List, Optional, Protocol


class Severity(str, Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    CRITICAL = "Critical"


@dataclass
class SecuritySignal:
    run_id: str
    user_hash: str
    event: str
    tool: str
    severity: Severity
    reason: str
    time: Optional[datetime] = None
    attrs: Dict[str, str] = field(default_factory=dict)


class CounterStore:
    def __init__(self) -> None:
        self._counts: Dict[str, int] = {}
        self._lock = threading.Lock()

    def inc(self, key: str) -> None:
        with self._lock:
            self._counts[key] = self._counts.get(key, 0) + 1

    def get(self, key: str) -> int:
        with self._lock:
            return self._counts.get(key, 0)


@dataclass
class Alert:
    name: str
    severity: Severity
    message: str
    run_id: str


class AlertSink(Protocol):
    def send(self, alert: Alert) -> None: ...


@dataclass
class Rule:
    name: str
    event: str
    threshold: int
    severity: Severity

    def evaluate(
        self, signal: SecuritySignal, store: CounterStore
    ) -> Optional[Alert]:
        if signal.event != self.event:
            return None

        key = f"{self.event}:{signal.user_hash}:{signal.tool}"
        store.inc(key)

        if store.get(key) < self.threshold:
            return None

        return Alert(
            name=self.name,
            severity=self.severity,
            message=(
                f"threshold reached for event={signal.event} tool={signal.tool}"
            ),
            run_id=signal.run_id,
        )


@dataclass
class Monitor:
    store: CounterStore
    rules: List[Rule]
    sink: AlertSink

    def handle(self, signal: SecuritySignal) -> None:
        if signal.time is None:
            signal.time = datetime.now(timezone.utc)

        for rule in self.rules:
            alert = rule.evaluate(signal, self.store)
            if alert is None:
                continue
            self.sink.send(alert)


RULES: List[Rule] = [
    Rule(
        name="Prompt injection spike",
        event="prompt_injection_detected",
        threshold=5,
        severity=Severity.HIGH,
    ),
    Rule(
        name="Repeated denied tool calls",
        event="tool_denied",
        threshold=3,
        severity=Severity.MEDIUM,
    ),
    Rule(
        name="Blocked egress attempts",
        event="egress_blocked",
        threshold=1,
        severity=Severity.HIGH,
    ),
    Rule(
        name="Budget runaway",
        event="budget_exceeded",
        threshold=1,
        severity=Severity.HIGH,
    ),
]


@dataclass
class WeakSignals:
    """Weak signals for one run (scope drift / monitoring tampering)."""

    new_external_domain: bool = False
    credential_search: bool = False
    monitoring_tampering: bool = False
    out_of_scope_action: bool = False
    continue_after_deny: bool = False
    real_secret_access: bool = False


def should_auto_stop(s: WeakSignals) -> bool:
    """True on emergency-stop hard trigger. Stop via §17.

    Correlation (new domain + credential search + tampering + out-of-scope)
    is covered by these flags; credential_search alone does not stop.
    """
    return (
        s.monitoring_tampering
        or s.continue_after_deny
        or s.real_secret_access
        or s.new_external_domain
        or s.out_of_scope_action
    )


@dataclass
class TrajectorySignals:
    """Compositional flags for one run (≠ scope_drift)."""

    external_identity_created: bool = False
    unsolicited_human_contact: bool = False
    unrequested_repo_modify: bool = False
    goal_read_only_analysis: bool = False


def should_stop_trajectory(t: TrajectorySignals) -> bool:
    """True on compositional violation vs goal. Separate from should_auto_stop."""
    if t.external_identity_created and t.unsolicited_human_contact:
        return True
    if t.goal_read_only_analysis and t.unrequested_repo_modify:
        return True
    return False

File: examples/typescript/part-5/16-monitoring-alerting.ts
// Illustrative examples for notes/part-5-control-observability/16-monitoring-alerting.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

enum Severity {
  Low = "Low",
  Medium = "Medium",
  High = "High",
  Critical = "Critical",
}

interface SecuritySignal {
  time?: Date;
  runId: string;
  userHash: string;
  event: string;
  tool: string;
  severity: Severity;
  reason: string;
  attrs: Record<string, string>;
}

class CounterStore {
  private readonly counts = new Map<string, number>();

  inc(key: string): void {
    this.counts.set(key, (this.counts.get(key) ?? 0) + 1);
  }

  get(key: string): number {
    return this.counts.get(key) ?? 0;
  }
}

interface Alert {
  name: string;
  severity: Severity;
  message: string;
  runId: string;
}

interface AlertSink {
  send(alert: Alert): void;
}

class Rule {
  name: string;
  event: string;
  threshold: number;
  severity: Severity;

  constructor(name: string, event: string, threshold: number, severity: Severity) {
    this.name = name;
    this.event = event;
    this.threshold = threshold;
    this.severity = severity;
  }

  evaluate(signal: SecuritySignal, store: CounterStore): Alert | undefined {
    if (signal.event !== this.event) {
      return undefined;
    }

    const key = `${this.event}:${signal.userHash}:${signal.tool}`;
    store.inc(key);

    if (store.get(key) < this.threshold) {
      return undefined;
    }

    return {
      name: this.name,
      severity: this.severity,
      message: `threshold reached for event=${signal.event} tool=${signal.tool}`,
      runId: signal.runId,
    };
  }
}

class Monitor {
  store: CounterStore;
  rules: Rule[];
  sink: AlertSink;

  constructor(store: CounterStore, rules: Rule[], sink: AlertSink) {
    this.store = store;
    this.rules = rules;
    this.sink = sink;
  }

  handle(signal: SecuritySignal): void {
    if (!signal.time) {
      signal.time = new Date();
    }

    for (const rule of this.rules) {
      const alert = rule.evaluate(signal, this.store);
      if (!alert) {
        continue;
      }
      this.sink.send(alert);
    }
  }
}

const RULES: Rule[] = [
  new Rule("Prompt injection spike", "prompt_injection_detected", 5, Severity.High),
  new Rule("Repeated denied tool calls", "tool_denied", 3, Severity.Medium),
  new Rule("Blocked egress attempts", "egress_blocked", 1, Severity.High),
  new Rule("Budget runaway", "budget_exceeded", 1, Severity.High),
];

/** Weak signals for one run (scope drift / monitoring tampering). */
interface WeakSignals {
  newExternalDomain?: boolean;
  credentialSearch?: boolean;
  monitoringTampering?: boolean;
  outOfScopeAction?: boolean;
  continueAfterDeny?: boolean;
  realSecretAccess?: boolean;
}

/**
 * True on emergency-stop hard trigger. Stop via §17.
 * Correlation (new domain + credential search + tampering + out-of-scope)
 * is covered by these flags; credentialSearch alone does not stop.
 */
function shouldAutoStop(s: WeakSignals): boolean {
  return Boolean(
    s.monitoringTampering ||
      s.continueAfterDeny ||
      s.realSecretAccess ||
      s.newExternalDomain ||
      s.outOfScopeAction,
  );
}

export { shouldAutoStop, type WeakSignals };
